lib/utils.js

```diff
@@ -4,3 +4,105 @@ export function PDFNumber(n) {
   // @see ISO 32000-1 Annex C.2 (real numbers)
   return Math.fround(n);
 }
+
+/**
+ * Measurement of size
+ *
+ * @typedef {number | `${number}` | `${number}${'em' | 'in' | 'px' | 'cm' | 'mm' | 'pc' | 'ex' | 'ch' | 'rem' | 'vw' | 'vmin' | 'vmax' | '%' | 'pt'}`} Size
+ */
+
+/**
+ * Measurement of how wide something is, false means 0 and true means 1
+ *
+ * @typedef {Size | boolean} Wideness
+ */
+
+/**
+ * Side definitions
+ * - To define all sides, use a single value
+ * - To define up-down left-right, use a `[Y, X]` array
+ * - To define each side, use `[top, right, bottom, left]` array
+ * - Or `{vertical: SideValue, horizontal: SideValue}`
+ * - Or `{top: SideValue, right: SideValue, bottom: SideValue, left: SideValue}`
+ *
+ * @template T
+ * @typedef {T| [T, T]| [T, T, T, T]| { vertical: T; horizontal: T }| { top: T; right: T; bottom: T; left: T }} SideDefinition<T>
+ **/
+
+/**
+ * @template T
+ * @typedef {{ top: T; right: T; bottom: T; left: T }} ExpandedSideDefinition<T>
+ */
+
+/**
+ * Convert any side definition into a static structure
+ *
+ * @template S
+ * @template D
+ * @template O
+ * @template {S | D} T
+ * @param {SideDefinition<S>} sides - The sides to convert
+ * @param {SideDefinition<D>} defaultDefinition - The value to use when no definition is provided
+ * @param {function(T): O} transformer - The transformation to apply to the sides once normalized
+ * @returns {ExpandedSideDefinition<O>}
+ */
+export function normalizeSides(
+  sides,
+  defaultDefinition = undefined,
+  transformer = (v) => v,
+) {
+  if (
+    sides === undefined ||
+    (typeof sides === "object" && Object.keys(sides).length === 0)
+  ) {
+    sides = defaultDefinition;
+  }
+  if (typeof sides !== "object" || sides === null || sides === undefined) {
+    sides = [sides, sides, sides, sides];
+  }
+  if (Array.isArray(sides)) {
+    if (sides.length === 2) {
+      sides = { vertical: sides[0], horizontal: sides[1] };
+    } else {
+      sides = {
+        top: sides[0],
+        right: sides[1],
+        bottom: sides[2],
+        left: sides[3],
+      };
+    }
+  }
+
+  if ("vertical" in sides || "horizontal" in sides) {
+    sides = {
+      top: sides.vertical,
+      right: sides.horizontal,
+      bottom: sides.vertical,
+      left: sides.horizontal,
+    };
+  }
+
+  if (
+    !(
+      "top" in sides ||
+      "right" in sides ||
+      "bottom" in sides ||
+      "left" in sides
+    )
```

Comment:
I believe {top: 1, right: 1} should get invalid value: {top: 1, right: 1, left: {top: 1, right: 1}, right: {top: 1, right: 1}}

Or do i miss something?

Reply:
I would remove this check at all assuming is undefined is treated as 0. If i do {top: 1, right: 1} i expect left and bottom be treated as 0

Reply:
the idea is to allow partials, so if anyone provides at least one of the keys it will be used and the others set to undefined, however if none of the keys match the object itself will be treated as the value for all keys

Reply:
other option we could take is to just say if its an object and you didnt provide the keys then it'll be undefined

Reply:
thinking about it now, it really doesnt make any sense to provide this, its the users fault

Reply:
@blikblum please see new changes which removes this logic